Admin documentation (#633)
SwissDataScienceCenter/renku: docs/admin/index.rst

       - `Setup Kubernetes on Openstack <prerequisites/k8s/openstack.html>`_
       - You can also `Setup a Kubernetes cluser with Rancher <prerequisites/k8s/rancher.html>`_
   - a `loadbalancer <prerequisites/loadbalancer.html>`_
   - a `Gitlab runner <prerequisites/gitlabrunner.html>`_ VM for the docker image builder with **sufficient storage**
   - a setup of `Helm <prerequisites/tiller.html>`_
   - a way to obtain `TLS certificates <prerequisites/certificates.html>`_
   - an `NGINX Ingress Controller  <prerequisites/nginx.html>`_ configured
   - a `DNS <prerequisites/dns.html>`_ domain for Renku

Optionally, you can choose to manage your own instances of the following:

   - a `Gitlab stand alone instance <https://about.gitlab.com/install/>`_
   - a Keycloak stand alone instance

B. Pre-deployment steps
-----------------------

1. (optional) Stand alone configurations
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

   - `Stand alone Gitlab <configurations/standalone-gitlab.html>`_
   - `Stand alone Keycloak <configurations/standalone-keycloak.html>`_

2. Create and configure Renku PVs and PVCs
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> All of Renku information is stored in three volumes: Jena, Postgresql and Gitlab (if not stand alone).

Reviewer: ```suggestion
All of Renku information is stored in three volumes needed by Jena, Postgresql and Gitlab (if not stand-alone).
```

To this end, persistent volumes and persistent volume claims need to be created. You can use the following yaml files as a base to execute the `kubectl` commands below.

   - `Renku PV yaml file <https://github.com/SwissDataScienceCenter/renku-admin-docs/blob/master/renku-pv.yaml>`_
   - `Renku PVC yaml file <https://github.com/SwissDataScienceCenter/renku-admin-docs/blob/master/renku-pvc.yaml>`_

   .. code-block:: console

     $ kubectl create -f renku-pv.yaml
     $ kubectl create ns renku
     $ kubectl -n renku create -f renku-pvc.yaml

>      $ kubectl -n renku edit pvc renku-postgresql

Reviewer: what are you editing?

     $ kubectl get pv
     $ kubectl get pvc -n renku
     $ kubectl describe persistentvolumeclaim -n renku  ## this should not show any error, just PVs ready to be used

> Ensure that PVs have a `default storageclass` and the following value is set:

Reviewer: ```suggestion
Ensure that PVs have the appropriate storageclass and the following value is set:
```

   .. code-block:: console

     metadata:
       name: default
       annotations:

>          storageclass.beta.kubernetes.io/is-default-class: "true"

Reviewer: why? someone might not want this - very cluster-specific

3. Create a renku-values.yaml file
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This step is very important as it will define the values that helm will use to install your Renku instance.

> You can find a basic file in `Renku values file <https://github.com/SwissDataScienceCenter/renku-admin-docs/blob/master/renku-values.yaml>`_

Reviewer: I wouldn't point them here - that file is not documented at all - just start with the relatively well-documented values file that is bundled with the helm chart in the renku repo? 

We should think about adapting the script from the dev-docs repo for this...

Author: I actually did document a bit better the one that's there , hence the pending PR.
About the script, exactly my thoughts, I guess we can do that for a second admin docs version.

Ensure the following has been taken care of before deploying Renku:

  - `variables_switchboard` section has URLs and DNS names that correspond to your configuration
  - `credentials` section has correctly populated values. Please see the comments in the renku-values file for more information.
  - check the `resources.requests.memory` value makes sense
  - `lfsObjects` & `registry` require to setup your S3 back-ends, so you need to have that configuration handy and in place

C. Deploying Renku
------------------

> 1. (Optional step for standalone Gitlab)

Reviewer: ```suggestion
1. (Optional step for stand-alone Gitlab)
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> To be able to support notebooks in private projects we need to insert the gitlab sudo token as a secret so that docker can download the corresponding image from the registry.

Reviewer: ```suggestion
To be able to support notebooks in private projects we need to insert the gitlab admin token as a secret so that docker can download the corresponding private images from the registry.
```

it is _not_ a sudo token - it only needs to have admin registry privileges - a sudo token implies that it can impersonate any user for the API

.. code-block:: bash

  $ kubectl -n renku create secret docker-registry renku-notebooks-registry \\
  --docker-server=<registryURL>:<port> --docker-username=root \\
  --docker-password=<gitlabSudoToken> --docker-email=root@renku-mydomain

2. (Optional certificates)
~~~~~~~~~~~~~~~~~~~~~~~~~~

> If you chose to create a certificate instead of using LetsEncrypt or similar, you can create the TLS secret with the following command:

Reviewer: ```suggestion
If you chose to create a certificate manually instead of using LetsEncrypt or similar, you can create the TLS secret with the following command:
```

.. code-block:: bash

   $ kubectl -n renku create secret tls renku-mydomain-ch-tls --cert=certificate.crt --key=certificate.key

> Note that renku-mydomain-ch-tls should correspond to the `ingress TLS value in Renku values file <https://github.com/SwissDataScienceCenter/renku-admin-docs/blob/master/renku-values.yaml#L12>`_

Reviewer: ```suggestion
Note that ``renku-mydomain-ch-tls`` should correspond to the `ingress TLS value in Renku values file <https://github.com/SwissDataScienceCenter/renku-admin-docs/blob/master/renku-values.yaml#L12>`_
```

3. Deploy renku
~~~~~~~~~~~~~~~

> To deploy Renku you can use the following command:

Reviewer: ```suggestion
Once all the pieces are in place, you can deploy Renku with the following commands:
```

.. code-block:: console

    $ helm init
    $ helm repo add renku https://swissdatasciencecenter.github.io/helm-charts/
    $ helm upgrade --install renku renku/renku \
     --namespace renku \

>      --version $(cat renku-version.txt) \

Reviewer: ```suggestion
     --version <renku-version> \
```

     -f renku-values.yaml \
     --timeout 1800

During deployment you can check the Renku pods being started.

4. Post deployment configuration
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

After Renku has been deployed you can make some post deployment configurations.
For instance, make a user admin on gitlab.

> 1. turn off automatic redirect to gitlab by setting gitlab.oauth.autoSignIn: false

Reviewer: ```suggestion
1. turn off automatic redirect to gitlab by setting redeploying with the value ``gitlab.oauth.autoSignIn: false``
```

> 2. log in as the root user using the password from gitlab.password

Reviewer: ```suggestion
2. log in as the root user using the password from ``gitlab.password``
```

3. modify any users you want to modify (e.g. to make them admin)
4. turn the automatic redirect back on

> D. Verifying Renku

Reviewer: ```suggestion
Verifying Renku
```
